examples/test_messaging/test_messaging.c

```diff
@@ -0,0 +1,398 @@
+                        printf("---------------------------\n");
+                        msg_params->tests_failed++;
+                        msg_params->test_phase++;
+                }
+                else {
+                        msg_params->test_msg_count++;
+                }
+
+                rte_free(msg_data);
+
+                if(10 == msg_params->test_msg_count){
+
+                        if(rte_ring_count(msg_params->msg_q) != 0) {
+                                printf("FAILED TEST 2: Shouldn't be any messages left, but there are %d in the ring. This may cause future tests to fail.\n", rte_ring_count(msg_params->msg_q));
+                                printf("---------------------------\n");
+                                while(rte_ring_count(msg_params->msg_q) != 0){
+                                        rte_ring_dequeue(msg_params->msg_q, msg_data);
+                                }
+                                msg_params->tests_failed++;
+                                msg_params->test_phase++;
+                        }
+                        else {
+                                printf("PASSED TEST 2\n");
+                                printf("---------------------------\n");
+                                msg_params->tests_passed++;
+                                msg_params->test_phase++;
+                                msg_params->test_msg_count = 0;
+                        }
+
+                }
+                
+                if(msg_params->test_msg_count == 0){
+                        printf("TEST 3: Message Ring Overflow...\n");
+                        printf("---------------------------\n");
+                        for(int i = 0; i < 130; i++){
+                                int* msg_int = (int*)rte_malloc(NULL, sizeof(int), 0);
+                                *msg_int = i;    
+                                onvm_nflib_send_msg_to_nf(msg_params->address, msg_int);
+                        }
+                }
+
+                // TODO: can we detect that we fail phase 2 from missing messages?
+
+        }
+        //Tests to see if even with a ring overflow it can still handle the messages
+        else if(msg_params->test_phase == 3){
+                
+                if(*((int *)msg_data) != msg_params->test_msg_count) { 
+                        printf("FAILED TEST 3: received %d instead of %d\n", *((int *)msg_data), msg_params->test_msg_count);
+                        printf("---------------------------\n");
+                        msg_params->tests_failed++;
+                        msg_params->test_phase++;
+                }
+                else {
+                        msg_params->test_msg_count++;
+                }
+
+                rte_free(msg_data);
+                
+                if(127 == msg_params->test_msg_count){
+
+                        if(rte_ring_count(msg_params->msg_q) != 0) {
```

Comment:
Currently we can only fail if there are extra messages in the ring (which should be impossible, unless we didn't get the right ring size).

This won't be able to catch the original memory leak error. The memory leak would cause messages to be created in the `nf_msg_pool` (with `rte_mempool_get`) but they would never be freed (using the `rte_mempool_put` command. I think the only way to detect if that is happening is to use one of these functions:

```
//https://doc.dpdk.org/api/rte__mempool_8h.html#a39ca98d81d5a0cdcf4fdf38f42bebb67
unsigned int | rte_mempool_avail_count (const struct rte_mempool *mp)
unsigned int | rte_mempool_in_use_count (const struct rte_mempool *mp)
static int | rte_mempool_full (const struct rte_mempool *mp)
static int | rte_mempool_empty (const struct rte_mempool *mp)
```
After all messages have been processed, the memory pool should be empty (or maybe the message you are processing will still be allocated, I'm not sure when it gets freed).  If there are still multiple messages allocated from the pool, but none in our queue, then we know that there must be a memory leak.

(Of course all of this assumes that no other NFs are running at the same time and sending messages, but I think that is a safe assumption since this is a testing NF)